File: ai-worker/src/routes/plan.py
"""Plan generation endpoints."""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional
from datetime import date

from src.planner.plan_generator import PlanGenerator

router = APIRouter()

plan_generator = PlanGenerator()


class PlanRequest(BaseModel):
    """Request for plan generation."""
    children_ages: list[int]
    date: date
    budget: Optional[float] = 50.0
    lat: Optional[float] = 49.0069  # Karlsruhe
    lng: Optional[float] = 8.4037
    preferences: Optional[dict] = None


class TimeSlot(BaseModel):
    """A time slot in the plan."""
    event_id: Optional[str]
    event_title: Optional[str]
    slot_type: str  # activity, break, travel
    start_time: str
    end_time: str
    duration_minutes: int
    notes: Optional[str]


class PlanResponse(BaseModel):
    """Generated plan response."""
    date: str
    children_ages: list[int]
    budget: float
    estimated_cost: float
    main_plan: list[TimeSlot]
    plan_b: list[TimeSlot]
    tips: list[str]
    generated_at: str


@router.post("/generate", response_model=PlanResponse)
async def generate_plan(request: PlanRequest):
    """
    Generate a family day plan using AI.
    
    Takes into account:
    - Children's ages for suitable activities
    - Budget constraints
    - Location for route optimization
    - Preferences (indoor/outdoor, categories)
    
    Returns a main plan and a Plan B for bad weather.
    """
    try:
        result = await plan_generator.generate(
            children_ages=request.children_ages,
            target_date=request.date,
            budget=request.budget or 50.0,
            lat=request.lat or 49.0069,
            lng=request.lng or 8.4037,
            preferences=request.preferences or {}
        )
        
        return PlanResponse(
            date=str(request.date),
            children_ages=request.children_ages,
            budget=request.budget or 50.0,
            estimated_cost=result.estimated_cost,
            main_plan=[TimeSlot(**slot) for slot in result.main_slots],
            plan_b=[TimeSlot(**slot) for slot in result.plan_b_slots],
            tips=result.tips,
            generated_at=result.generated_at
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Plan generation failed: {str(e)}")


class OptimizeRequest(BaseModel):
    """Request for plan optimization."""
    events: list[dict]
    children_ages: list[int]
    date: str
    budget: float = 50.0
    preferences: Optional[dict] = None
    user_location: Optional[dict] = None


@router.post("/optimize")
async def optimize_plan(request: OptimizeRequest):
    """
    Select and optimize events for a family day plan.
    
    Uses scoring and heuristics to select the best combination of events.
    Returns selected event IDs for main plan and Plan B.
    """
    events = request.events
    children_ages = request.children_ages
    budget = request.budget
    preferences = request.preferences or {}
    
    if not events:
        return {
            "selected_events": [],
            "plan_b_events": [],
            "reasoning": "No events available"
        }
    
    # Score events based on multiple factors
    scored_events = []
    for event in events:
        score = 0.0
        
        # Base score from AI scoring (if available)
        if event.get('scores'):
            score += (event['scores'].get('family_fit_score', 50) or 50) * 0.3
            score += (event['scores'].get('stressfree_score', 50) or 50) * 0.2
            score += (event['scores'].get('quality_score', 50) or 50) * 0.1
        else:
            score += 50  # Default score
        
        # Distance penalty
        distance = event.get('distance_km', 10)
        score -= distance * 2  # -2 points per km
        
        # Price bonus for free events
        if event.get('price_type') == 'free':
            score += 15
        elif event.get('price_min') and float(event['price_min']) <= budget / 4:
            score += 10
        
        # Age suitability bonus
        age_range = (event.get('age_min', 0), event.get('age_max', 18))
        ages_suitable = sum(1 for age in children_ages if age_range[0] <= age <= age_range[1])
        score += ages_suitable * 5
        
        # Category diversity tracking
        categories = event.get('categories', [])
        
        scored_events.append({
            **event,
            'computed_score': score,
            'categories': categories
        })
    
    # Sort by score
    scored_events.sort(key=lambda e: e['computed_score'], reverse=True)
    
    # Select events with diversity in mind
    selected = []
    used_categories = set()
    total_cost = 0
    
    for event in scored_events:
        if len(selected) >= 3:
            break
        
        # Check budget
        event_cost = 0
        if event.get('price_type') != 'free' and event.get('price_min'):
            event_cost = float(event['price_min']) * (len(children_ages) + 2)
        
        if total_cost + event_cost > budget:
            continue
        
        # Prefer category diversity for first 2 events
        event_cats = set(event.get('categories', []))
        if len(selected) < 2:
            has_new_cat = not event_cats or not event_cats.issubset(used_categories)
            if not has_new_cat and len(scored_events) > len(selected) + 3:
                continue  # Skip if no new category and more options available
        
        selected.append(event['id'])
        used_categories.update(event_cats)
        total_cost += event_cost
    
    # Select Plan B events (indoor alternatives)
    plan_b = []
    for event in scored_events:
        if event['id'] in selected:
            continue
        if not event.get('is_indoor'):
            continue
        if len(plan_b) >= 3:
            break
        plan_b.append(event['id'])
    
    return {
        "selected_events": selected,
        "plan_b_events": plan_b,
        "reasoning": f"Selected {len(selected)} events optimized for family fit, distance, and budget. Plan B has {len(plan_b)} indoor alternatives."
    }


@router.post("/optimize-route")
async def optimize_route(slots: list[dict], start_location: dict):
    """
    Optimize the route between plan slots.
    
    Returns optimized order and travel times.
    """
    if not slots:
        return {
            "optimized_order": [],
            "total_travel_time_minutes": 0
        }
    
    # Simple nearest-neighbor optimization
    # For a full implementation, use Google Maps Directions API
    optimized_order = list(range(len(slots)))
    
    # Estimate travel time based on distance (rough: 3 min per km)
    total_travel = 0
    for i, slot in enumerate(slots):
        if i > 0:
            # Assume average 5km between locations
            total_travel += 15  # 15 min average
    
    return {
        "optimized_order": optimized_order,
        "total_travel_time_minutes": total_travel,
        "note": "Basic optimization. For better results, integrate Google Maps Directions API."
    }

File: backend/src/routes/plan.ts
import { Router, Request, Response, NextFunction } from 'express';
import { body, validationResult } from 'express-validator';
import { prisma } from '../lib/prisma.js';
import { createError } from '../middleware/errorHandler.js';
import { requireAuth, optionalAuth, type AuthRequest } from '../middleware/auth.js';
import { searchEventsWithinRadius } from '../lib/geo.js';
import { logger } from '../lib/logger.js';

const router = Router();

const AI_WORKER_URL = process.env.AI_WORKER_URL || 'http://localhost:5000';

// Validation for plan generation
const validatePlanRequest = [
  body('children_ages').isArray({ min: 1 }).withMessage('At least one child age required'),
  body('children_ages.*').isInt({ min: 0, max: 18 }),
  body('date').isISO8601(),
  body('budget').optional().isFloat({ min: 0 }),
  body('lat').optional().isFloat({ min: -90, max: 90 }),
  body('lng').optional().isFloat({ min: -180, max: 180 }),
  body('preferences').optional().isObject(),
];

// POST /api/plan/generate - Generate a family day plan
router.post('/generate', validatePlanRequest, async (req: Request, res: Response, next: NextFunction) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      throw createError('Validation error', 400, 'VALIDATION_ERROR');
    }

    const {
      children_ages,
      date,
      budget = 50,
      lat = 49.0069, // Karlsruhe center
      lng = 8.4037,
      radius = 15, // km
      preferences = {},
      use_ai = false, // Optional: use AI for plan optimization
    } = req.body;

    // Determine age range
    const minAge = Math.min(...children_ages);
    const maxAge = Math.max(...children_ages);

    // Build query for suitable events
    const targetDate = new Date(date);
    const dayStart = new Date(targetDate);
    dayStart.setHours(8, 0, 0, 0);
    const dayEnd = new Date(targetDate);
    dayEnd.setHours(20, 0, 0, 0);

    // Use geo-search to find events within radius
    let availableEvents: any[] = [];
    
    try {
      const geoResult = await searchEventsWithinRadius(lat, lng, radius, {
        limit: 30,
        status: 'published',
        dateFrom: dayStart,
        dateTo: dayEnd,
      });
      
      // Fetch full event data with categories and scores
      if (geoResult.events.length > 0) {
        const eventIds = geoResult.events.map(e => e.id);
        availableEvents = await prisma.canonicalEvent.findMany({
          where: {
            id: { in: eventIds },
            is_complete: true,
            // Age filter
            OR: [
              { age_min: null, age_max: null }, // No age restriction
              {
                AND: [
                  { OR: [{ age_min: null }, { age_min: { lte: maxAge } }] },
                  { OR: [{ age_max: null }, { age_max: { gte: minAge } }] },
                ]
              }
            ]
          },
          include: {
            categories: { include: { category: true } },
            scores: true,
          },
        });
        
        // Add distance from geo search
        availableEvents = availableEvents.map(event => {
          const geoEvent = geoResult.events.find(e => e.id === event.id);
          return { ...event, distance_km: geoEvent?.distance_km };
        });
      }
    } catch (geoError) {
      logger.warn('Geo search failed, falling back to standard query', { error: geoError });
      
      // Fallback to standard query without geo
      const where: any = {
        status: 'published',
        is_complete: true,
        is_cancelled: false,
        start_datetime: { gte: dayStart, lte: dayEnd },
        OR: [
          { age_min: null, age_max: null },
          {
            AND: [
              { OR: [{ age_min: null }, { age_min: { lte: maxAge } }] },
              { OR: [{ age_max: null }, { age_max: { gte: minAge } }] },
            ]
          }
        ]
      };
      
      availableEvents = await prisma.canonicalEvent.findMany({
        where,
        include: {
          categories: { include: { category: true } },
          scores: true,
        },
        orderBy: [
          { scores: { family_fit_score: 'desc' } },
          { scores: { stressfree_score: 'desc' } },
        ],
        take: 30,
      });
    }

    // Apply budget filter
    if (budget <= 0) {
      availableEvents = availableEvents.filter(e => e.price_type === 'free');
    } else {
      availableEvents = availableEvents.filter(e => 
        e.price_type === 'free' || 
        (e.price_min && Number(e.price_min) <= budget / 2)
      );
    }

    // Apply preferences
    if (preferences.indoor === true && preferences.outdoor !== true) {
      availableEvents = availableEvents.filter(e => e.is_indoor);
    } else if (preferences.outdoor === true && preferences.indoor !== true) {
      availableEvents = availableEvents.filter(e => e.is_outdoor);
    }

    if (preferences.categories?.length > 0) {
      availableEvents = availableEvents.filter(e => 
        e.categories?.some((c: any) => preferences.categories.includes(c.category?.slug))
      );
    }

    // Sort by score and distance
    availableEvents.sort((a, b) => {
      const scoreA = (a.scores?.family_fit_score || 0) + (a.scores?.stressfree_score || 0);
      const scoreB = (b.scores?.family_fit_score || 0) + (b.scores?.stressfree_score || 0);
      const distanceA = a.distance_km || 100;
      const distanceB = b.distance_km || 100;
      
      // Weighted score: quality (60%) + proximity (40%)
      const weightedA = scoreA * 0.6 - distanceA * 4;
      const weightedB = scoreB * 0.6 - distanceB * 4;
      
      return weightedB - weightedA;
    });

    // Select events for the plan
    let selectedEvents: any[];
    let planBEvents: any[];
    let aiUsed = false;

    // Try AI-based selection if requested and available
    if (use_ai && AI_WORKER_URL) {
      try {
        const aiResponse = await fetch(`${AI_WORKER_URL}/plan/optimize`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            events: availableEvents.map(e => ({
              id: e.id,
              title: e.title,
              description: e.description_short,
              start_datetime: e.start_datetime,
              end_datetime: e.end_datetime,
              location_address: e.location_address,
              location_lat: e.location_lat,
              location_lng: e.location_lng,
              price_type: e.price_type,
              price_min: e.price_min,
              is_indoor: e.is_indoor,
              is_outdoor: e.is_outdoor,
              categories: e.categories?.map((c: any) => c.category?.slug),
              scores: e.scores,
              distance_km: e.distance_km,
            })),
            children_ages,
            date: targetDate.toISOString(),
            budget,
            preferences,
            user_location: { lat, lng },
          }),
          signal: AbortSignal.timeout(10000), // 10 second timeout
        });

        if (aiResponse.ok) {
          const aiResult = await aiResponse.json();
          if (aiResult.selected_events && aiResult.plan_b_events) {
            selectedEvents = aiResult.selected_events.map((id: string) => 
              availableEvents.find(e => e.id === id)
            ).filter(Boolean);
            planBEvents = aiResult.plan_b_events.map((id: string) => 
              availableEvents.find(e => e.id === id)
            ).filter(Boolean);
            aiUsed = true;
          }
        }
      } catch (aiError) {
        logger.warn('AI plan optimization failed, using fallback', { error: aiError });
      }
    }

    // Fallback to rule-based selection
    if (!aiUsed) {
      selectedEvents = selectEventsForPlan(availableEvents, {
        targetCount: 3,
        children_ages,
      });

      // Get Plan B (indoor alternatives)
      planBEvents = availableEvents
        .filter(e => e.is_indoor && !selectedEvents.includes(e))
        .slice(0, 3);
    }

    // Create time slots
    const slots = createTimeSlots(selectedEvents, targetDate);
    const planBSlots = createTimeSlots(planBEvents, targetDate);

    // Calculate total cost
    const estimatedCost = selectedEvents.reduce((sum, event) => {
      if (event.price_type === 'free') return sum;
      const price = Number(event.price_min) || 0;
      return sum + (price * (children_ages.length + 2)); // Assume 2 adults
    }, 0);

    // Calculate total distance
    const totalDistance = selectedEvents.reduce((sum, event) => {
      return sum + (event.distance_km || 0);
    }, 0);

    const plan = {
      date: targetDate.toISOString().split('T')[0],
      children_ages,
      budget,
      estimated_cost: Math.round(estimatedCost * 100) / 100,
      total_distance_km: Math.round(totalDistance * 10) / 10,
      main_plan: {
        slots,
        total_events: slots.filter(s => s.event_id).length,
      },
      plan_b: {
        reason: 'Bei Regen oder wenn ausgebucht',
        slots: planBSlots,
      },
      tips: generateTips(selectedEvents, children_ages),
      meta: {
        events_considered: availableEvents.length,
        ai_optimized: aiUsed,
        search_radius_km: radius,
      }
    };

    res.json({
      success: true,
      data: plan,
    });
  } catch (error) {
    next(error);
  }
});

// ============================================
// SAVE & SHARE PLANS
// ============================================

// POST /api/plan/save - Save a generated plan
router.post('/save', optionalAuth, [
  body('date').isISO8601(),
  body('children_ages').isArray(),
  body('budget').optional().isFloat({ min: 0 }),
  body('title').optional().isString().isLength({ max: 200 }),
  body('slots').isArray(),
  body('plan_b_slots').optional().isArray(),
], async (req: Request, res: Response, next: NextFunction) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      throw createError('Validation error', 400, 'VALIDATION_ERROR');
    }

    const authReq = req as AuthRequest;
    const userId = authReq.user?.sub || null;

    const {
      date,
      children_ages,
      budget,
      title,
      slots,
      plan_b_slots = [],
      preferences,
    } = req.body;

    // Create the plan
    const plan = await prisma.plan.create({
      data: {
        user_id: userId,
        title: title || `Kiezling-Tag am ${new Date(date).toLocaleDateString('de-DE')}`,
        date: new Date(date),
        children_ages,
        budget: budget || null,
        preferences: preferences || null,
      }
    });

    // Create main plan slots
    for (const slot of slots) {
      await prisma.planSlot.create({
        data: {
          plan_id: plan.id,
          event_id: slot.event_id || null,
          slot_type: slot.type === 'break' ? 'break' : 'activity',
          start_time: new Date(slot.start_time),
          end_time: new Date(slot.end_time),
          duration_minutes: slot.duration_minutes,
          notes: slot.notes || slot.suggestion || null,
        }
      });
    }

    // Create Plan B slots
    for (const slot of plan_b_slots) {
      await prisma.planSlot.create({
        data: {
          plan_id: plan.id,
          plan_b_for_id: plan.id,
          event_id: slot.event_id || null,
          slot_type: slot.type === 'break' ? 'break' : 'activity',
          start_time: new Date(slot.start_time),
          end_time: new Date(slot.end_time),
          duration_minutes: slot.duration_minutes,
          notes: slot.notes || null,
        }
      });
    }

    // Generate share URL
    const shareUrl = `/plan/${plan.id}`;

    res.status(201).json({
      success: true,
      message: 'Plan gespeichert',
      data: {
        id: plan.id,
        share_url: shareUrl,
      }
    });
  } catch (error) {
    next(error);
  }
});

// GET /api/plan/:id - Get saved plan
router.get('/:id', async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { id } = req.params;

    const plan = await prisma.plan.findUnique({
      where: { id },
      include: {
        slots: {
          where: { plan_b_for_id: null },
          include: {
            event: {
              select: {
                id: true,
                title: true,
                description_short: true,
                location_address: true,
                price_type: true,
                price_min: true,
                is_indoor: true,
                is_outdoor: true,
              }
            }
          },
          orderBy: { start_time: 'asc' }
        },
        plan_b_slots: {
          include: {
            event: {
              select: {
                id: true,
                title: true,
                description_short: true,
                location_address: true,
                price_type: true,
                price_min: true,
                is_indoor: true,
                is_outdoor: true,
              }
            }
          },
          orderBy: { start_time: 'asc' }
        }
      }
    });

    if (!plan) {
      throw createError('Plan not found', 404, 'NOT_FOUND');
    }

    res.json({
      success: true,
      data: plan,
    });
  } catch (error) {
    next(error);
  }
});

// DELETE /api/plan/:id - Delete own plan
router.delete('/:id', requireAuth, async (req: Request, res: Response, next: NextFunction) => {
  try {
    const authReq = req as AuthRequest;
    const { id } = req.params;

    const plan = await prisma.plan.findUnique({
      where: { id }
    });

    if (!plan) {
      throw createError('Plan not found', 404, 'NOT_FOUND');
    }

    // Only owner can delete (admin check could be added)
    if (plan.user_id !== authReq.user!.sub) {
      throw createError('Not authorized', 403, 'FORBIDDEN');
    }

    await prisma.plan.delete({
      where: { id }
    });

    res.json({
      success: true,
      message: 'Plan deleted'
    });
  } catch (error) {
    next(error);
  }
});

// Helper functions

function selectEventsForPlan(events: any[], options: { targetCount: number; children_ages: number[] }) {
  const { targetCount } = options;
  
  if (events.length === 0) return [];
  if (events.length <= targetCount) return events;

  // Simple selection: take top events by score, ensuring variety
  const selected: any[] = [];
  const usedCategories = new Set<string>();

  for (const event of events) {
    if (selected.length >= targetCount) break;

    // Try to get variety in categories
    const eventCategories = event.categories?.map((c: any) => c.category?.slug) || [];
    const hasNewCategory = eventCategories.some((cat: string) => !usedCategories.has(cat));

    if (selected.length < 2 || hasNewCategory) {
      selected.push(event);
      eventCategories.forEach((cat: string) => usedCategories.add(cat));
    }
  }

  // Fill remaining slots if needed
  while (selected.length < targetCount && selected.length < events.length) {
    const remaining = events.filter(e => !selected.includes(e));
    if (remaining.length > 0) {
      selected.push(remaining[0]);
    } else {
      break;
    }
  }

  return selected;
}

function createTimeSlots(events: any[], date: Date) {
  const slots: any[] = [];
  let currentTime = new Date(date);
  currentTime.setHours(10, 0, 0, 0); // Start at 10 AM

  for (const event of events) {
    const startTime = new Date(currentTime);
    const duration = 90; // Default 90 minutes per activity
    const endTime = new Date(startTime.getTime() + duration * 60000);

    slots.push({
      event_id: event.id,
      event: {
        id: event.id,
        title: event.title,
        description_short: event.description_short,
        location_address: event.location_address,
        price_type: event.price_type,
        price_min: event.price_min,
        categories: event.categories,
        is_indoor: event.is_indoor,
        is_outdoor: event.is_outdoor,
      },
      start_time: startTime.toISOString(),
      end_time: endTime.toISOString(),
      duration_minutes: duration,
    });

    // Add 30 min break between activities
    currentTime = new Date(endTime.getTime() + 30 * 60000);

    // Add lunch break after 2nd activity
    if (slots.length === 2) {
      slots.push({
        type: 'break',
        title: 'Mittagspause',
        start_time: currentTime.toISOString(),
        end_time: new Date(currentTime.getTime() + 60 * 60000).toISOString(),
        duration_minutes: 60,
        suggestion: 'Zeit für ein Mittagessen in der Nähe',
      });
      currentTime = new Date(currentTime.getTime() + 60 * 60000);
    }
  }

  return slots;
}

function generateTips(events: any[], childrenAges: number[]) {
  const tips: string[] = [];

  // General tips
  if (childrenAges.some(age => age < 3)) {
    tips.push('Denke an Wickeltasche und Snacks für die Kleinen');
  }

  if (events.some(e => e.is_outdoor)) {
    tips.push('Sonnencreme und Wasser nicht vergessen');
  }

  if (events.length > 2) {
    tips.push('Plant genug Pausen ein - Kinder brauchen Zeit zum Verarbeiten');
  }

  tips.push('Fahrt mit dem ÖPNV? Die KVV-App zeigt Live-Verbindungen');

  return tips;
}

export default router;
